### ISSUE TITLE
Cannot push nuget packages to BaGet behind an  IIS Reverse Proxy

### ISSUE THREAD

**Issue by marcBalta** [loic-sharma/BaGet#276]
### Describe the bug
I'm using BaGet inside a windows docker image. IIS 10 is used as a reverse proxy. 
This is the web.config file used to proxy queries with the pattern domain.de/nuget/ to the BaGet server which is running in a Windows Docher container on another machine with the IP 192.168.178.2.

'''
		 <rule name="nuget" stopProcessing="true">
                    <match url="nuget/(.*)" />
                    <action type="Rewrite" url="http://192.168.178.2:5555/{R:1}" />
                    <serverVariables>
                        <set name="HTTP_FORWARDED" value="for={REMOTE_ADDR};by={LOCAL_ADDR};host=&quot;{HTTP_HOST}&quot;;proto=&quot;https&quot;" />
			<set name="HTTP_X_FORWARDED_HOST" value="{HTTP_HOST}" />
			<set name="HTTP_X_FORWARDED_SCHEMA" value="https" />
			<set name="HTTP_X_FORWARDED_PROTO" value="https" />
                    </serverVariables>
                </rule>

This is my baget.env :
'''				
ApiKey=NUGET-SERVER-API-KEY
PathBase=/nuget
Storage__Type=FileSystem
Storage__Path=c:\packages
Database__Type=Sqlite
Database__ConnectionString=Data Source=C:\packages\baget.db
Search__Type=Database
'''

When Im running however the nuget push command it uses the internal IP address (the url of docker container) and not the one of the proxy.
'''
$ dotnet nuget push  -s https://ge-apps-dev.intranet.mckinsey.com/nuget/v3/index.json -k NUGET-SERVER-API-KEY McK.Toolkit.Core.Asp.1.0.0.nupkg
info : Pushing McK.Toolkit.Core.Asp.1.0.0.nupkg to 'https://192.168.178.2:5555/api/v2/package'...
info :   PUT https://192.168.178.2:5555/api/v2/package/
info : An error was encountered when fetching 'PUT https://192.168.178.2:5555/api/v2/package/'. The request will now be retried.
info : A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond
info :   PUT https://192.168.178.2:5555/api/v2/package/
info : An error was encountered when fetching 'PUT https://192.168.178.2:5555/api/v2/package/'. The request will now be retried.
info : A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond
info :   PUT https://192.168.178.2:5555/api/v2/package/
error: A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond
'''

The angular UI isnt running at all. Thats seems to be however the typpicall problem with the base element not set correctly.

**Comment by Roemer** [comment]
I got the same issue (with an nginx reverse proxy).

**Comment by marcBalta** [comment]
In my opionion the problem is that BaGet doesnt configure the Forward Header Engine to use also XForwardedHost:

```
  var fordwardedHeaderOptions = new ForwardedHeadersOptions
            {
                ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.XForwardedProto | ForwardedHeaders.XForwardedHost
            };
  app.UseForwardedHeaders(fordwardedHeaderOptions);
```

**Comment by Roemer** [comment]
Pitty. Is there any workaround possible?

**Comment by marcBalta** [comment]
The workaround could be maybe to rewrite the domain in the http repsonses. But I didnt try that. If you got that working im keen to hear about your solution.

**Comment by Roemer** [comment]
It seems that those headers were once added, see https://github.com/loic-sharma/BaGet/pull/65/files
I guess they were missed while some migration took place and now they're gone. I am currently testing if that would fix that issue.

**Comment by marcBalta** [comment]
Ok interesting. But the important header is missing which is XForwardedHost. This contains contains the reverse proxies domain.

**Comment by Roemer** [comment]
yes, basically
```csharp
services.Configure<ForwardedHeadersOptions>(options =>
            {
                options.ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.XForwardedProto;
            });
```
is missing.

**Comment by Roemer** [comment]
Ah, `ForwardedHeaders.XForwardedHost` should probably be added as well.

**Comment by Roemer** [comment]
Well that part seems to work now. Now I also need to figure out how to change the base-path (I am running it in /nuget/ subfolder) when running from a docker container.

**Comment by Roemer** [comment]
There is a `app.UsePathBase(options.PathBase);` in `Startup.cs` but I cannot figure out how that options object is filled.

**Comment by marcBalta** [comment]
Cool. Will it be part of the next release?  In IIS I used a rewrite rule to solve this problem. It seemed that this made the nuget protocol working. But I didnt came so far to say that for sure. If not the following could help:
```
  if (Configuration["URL_BASE_PATH"] != null)
            {
                app.Use(async (ctx, next) =>
                {
                    ctx.Request.PathBase = Configuration["URL_BASE_PATH"];// "/schlemihl";
                    await next.Invoke();
                });
            }
```

 For the angular app, however, there has to be done some work: 
* The index.html has to be adapted
* and the services have to be rewriten to generate the reverse proxy url + base path

**Comment by marcBalta** [comment]
Well actually for my asp.net core apps UseBapthBase didnt work.

**Comment by Roemer** [comment]
Adding 
```csharp
services.Configure<ForwardedHeadersOptions>(options =>
{
    options.ForwardedHeaders = ForwardedHeaders.XForwardedFor | ForwardedHeaders.XForwardedProto | ForwardedHeaders.XForwardedHost;
});
```
to IServiceCollectionExtensions.cs
and
```csharp
string basePath = Environment.GetEnvironmentVariable("ASPNETCORE_BASEPATH");
if (!string.IsNullOrEmpty(basePath))
{
    app.Use(async (context, next) =>
    {
        context.Request.PathBase = basePath;
        await next.Invoke();
    });
    app.UsePathBase(options.PathBase);
}
```
to Startup.cs

Seems to fix the api entirely when I then provide the environment variable `ASPNETCORE_BASEPATH` in the docker container. The React UI still does not work.

**Comment by marcBalta** [comment]
Yeah I know. Please mind https://github.com/loic-sharma/BaGet/issues/276#issuecomment-499067303

**Comment by Roemer** [comment]
Yup but I don't know React. It is easy to hardcode the base path into it but it should be something dynamic so it can be controlled when running the docker container.

**Comment by marcBalta** [comment]
And will your changes be part of the next release?

**Comment by Roemer** [comment]
That you need to ask the author of this project ;) I could create a PR for it. I pushed a docker container with my changes here: https://hub.docker.com/r/roemer/baget In case you want to test it as well.

**Comment by marcBalta** [comment]
A PR would be great. Thx!

**Comment by Roemer** [comment]
Would be great to have the frontend fixed as well. Let's see if @loic-sharma has some hints / comments.

**Comment by Roemer** [comment]
Just for your info, I switched over to Nexus as that one works seamlessly with subdomains and reverse proxies.

**Comment by marcBalta** [comment]
Thx for the hint. I already switched to Nuget.Server but didnt had the time to put it into docker and behind a reverse proxy. I think I will have a look at nexus.

**Comment by DoCode** [comment]
@roman, @marcBalta thanks for the hints! Probably we should solve the root problem here.

**Comment by Roemer** [comment]
Yes that would be great. The code discussed above fix the issue that the headers passed from the reverse proxy are not handled correctly. With that, it generally works behind a reverse proxy.
The second issue is that the application should be able to run in a subfolder. It seems to be easy to add a hardcoded subpath on build time but I have no idea how to do that on runtime. That needs some investigation.

**Comment by loic-sharma** [comment]
Hello, I apologize for the long response time! It seems like there's two problems:

1. BaGet doesn't support reverse proxies well. This is a complete blocker and should be fixed ASAP
2. BaGet's UI doesn't work well in such scenarios. This is annoying, but fixing this is more a "nice to have"

To fix the first problem, it seems that `ForwardedHeaders.XForwardedHost` option should be added to [this line](https://github.com/loic-sharma/BaGet/blob/master/src/BaGet.Core.Server/Configuration/ConfigureForwardedHeadersOptions.cs#L11). Adding this option has security implications (see [CVE-2018-0787](https://github.com/aspnet/Announcements/issues/295)), so we should add configs to BaGet to lock down forwarding. I suggest we add the following configs:

```json
{
  ...

  "ForwardHeaders": {
    "Include": ["XForwardedFor", "XForwardedProto", "XForwardedHost" ],

    "AllowedHosts": [ "*.contoso.com" ],
    "KnownNetworks": [ "10.0.0.1" ],
    "KnownProxies": [ "10.0.0.1" ],
  },

  ...
}
```

These would map to [their equivalent ASP.NET Core configurations](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/proxy-load-balancer?view=aspnetcore-2.2#forwarded-headers-middleware-options). Does this seem like a good approach?

/cc @marcBalta @Roemer @DoCode

**Comment by Roemer** [comment]
@loic-sharma Yes you're right, these are two separate issues. The `ForwardedHeaders.XForwardedHost` should fix the first issue.
The second is (in my opinion) not only just a nice to have. Usually when using a reverse proxy, you want to redirect multiple applications to different servers.  For example:
```
nginx/app1 => appserver/app1
nginx/jenkins => buildserver/jenkins
nginx/nexus => nexusserver/nexus
```
Without the possibility to host an application under a subpath, it is (as far as I know) not possible to use it in a normal reverse proxy scenario.

**Comment by loic-sharma** [comment]
I see, that makes sense. Thanks for the information

**Comment by szwenni** [comment]
Hi

If the topic is still up to anyone I added full PathBase support. Refer to [basePathSupport](https://github.com/szwenni/BaGet/tree/basePathSupport) in my fork.
It basically adds a new option to BaGetOptions called `HostUrl` which allows to specify the AllowedHosts value.
As second part it rewrites all values in react to match the `PathBase`. This is done with a placeholder as homepage value in package.json and in config.tsx. For development use you have to specify the values by manually.
For proxying with nginx keep in mind to set `X-Forwarded-Host` to `$host:$server_port` 

@loic-sharma I will create a pull request for that.

**Comment by chrisbecke** [comment]
This is an IIS configuration issue not a BaGet issue.
The first setting you need to check on your re-write server is the server level setting: system.webServer/proxy: preserveHostHeader. If this is set "True" then the request "Host" header will be passed through to the destination server. And you don't need to worry about X-Forwarded or any of that confusing misdirection.
If it is set to "False" then ARR will rewrite the Host header to reflect the host hame of the server is is redirecting the request to. i.e. 192.168.etc.
Usually not what you want.
One way to fix this is clearly at this level but as the system.webServer/proxy preserverHostHeader setting applies server wide that might have unintended side effects if the server is hosting other loads making use of the proxy.
In which case, a simple rewrite of the rewrite rules has you covered:

So, first add "HTTP_HOST" to the list of Allowed Server Variables, then ammend your rewrite rule to have a condition and server variable rule like this :- 

Here the condition is always processed before the rule is triggered, and captures the "Host" header as {C:1}, and then serverVariables are applied hopefully after ARR and Rewrite have done their other modifications, where we set it back to the original - passed in - value.

```xml
            <rules>
                <rule name="test" stopProcessing="true">
                    <match url="(.*)" />
                    <conditions>
                        <add input="{HTTP_HOST}" pattern="(.*)" />
                    </conditions>
                    <serverVariables>
                        <set name="HTTP_HOST" value="{C:1}" />
                    </serverVariables>
                    <action type="Rewrite" url="http://example.com/{R:1}" />
                </rule>
```

Again. No need to mess around with non standard X-Forwarded etc. headers.

**Comment by trgiangvp3** [comment]
> This is an IIS configuration issue not a BaGet issue.
This is the BaGet issue.
When you run a push command from CLI, `nuget.exe` will request for a config json file that contains all necessaries url.
It then cache that config file at `\AppData\Local\NuGet\v3-cache\<some random string>$_nuget.devmoba.com_v3_index.json` and use the settings in that file to continue the process (of pushing packages). Changing some IIS config will definitely not help you to change this file 's content.

So, my work around is manually change all the URLs in that file to `http://my.company.name` instead of `http://locahost:xxxx`
Hopefully the next release will let us change this by a setting line.

**Comment by evgenyvalavin** [comment]
+1

**Comment by evgenyvalavin** [comment]
Found a WA:
Use outbound rule rewrite:
            <outboundRules>
                <rule name="Rewire internal host IP" preCondition="" patternSyntax="ECMAScript">
                        <match filterByTags="None" pattern="https:\/\/IP:PORT\/" negate="false" />
                        <action type="Rewrite" value="https://example.com/" />
                </rule>
            </outboundRules>

**Comment by chrisbecke** [comment]
> This is the BaGet issue.
> ...
>  Changing some IIS config will definitely not help you to change this file 's content.

You are correct in that, if an api is returning fully qualified urls, then it does take on the task of supporting different base urls itself.

However, "some IIS config" defiantly will help you change that files content. Response rewrite rules perform regex based replacements on outgoing content for specifically the purpose of fixing urls in response documents.

**Comment by Ririshi** [comment]
For anyone finding this now, here's a fix if you're hosting BaGet on IIS, behind another reverse proxy.

Some context to understand what's happening:
In short, BaGet generates its download NuGet links using the `Host` HTTP header, which is a server variable called `HTTP_HOST` in IIS. Behind a reverse proxy, this is set to the hostname or IP address that is used by the proxy server to connect to your BaGet server. The reverse proxy should set `HTTP_X_FORWARDED_HOST`, but BaGet does not listen to this, and will simply use `HTTP_HOST`. You need to either hard-code the `HTTP_HOST` variable to be the correct external hostname/IP address, or dynamically set its value to the value of `HTTP_X_FORWARDED_HOST`.

Steps to fix it if BaGet is internally hosted on IIS:
1. First, note that the urls in `https://your.baget.hostname/v3/index.json` are wrong, i.e. they contain the internal hostname/IP address of your BaGet server.
2. Download and install the [URL Rewrite](https://www.iis.net/downloads/microsoft/url-rewrite) module to your IIS machine.
3. Open IIS Manager, go to your site's home page, and open URL Rewrite.
5. Click "View Server variables" in the right side menu.
6. Add `HTTP_HOST` as an allowed variable from the right side menu.
7. Go back to the URL Rewrite page.
8. Add a blank rewrite rule from the right side menu.
9. Set the regex pattern to `(.*)` so it applies to all requests on this site.
10. Open the "Server Variables" collapsed menu and add `HTTP_HOST`.
11. Set the value to `{HTTP_X_FORWARDED_HOST}` and leave "replace the existing value" ticked.
12. In the "Action" collapsible menu, select the "None" action type to prevent actual URL rewriting.
13. Go back to the site main page and restart the site.
14. Refresh the URL from step 1, the contents should be correct now.

Note 1: if your reverse proxy does not set `HTTP_X_FORWARDED_HOST`, you can hard-code the value of `HTTP_HOST` to have the correct external hostname instead of the value used in step 8.

Note 2: if your reverse proxy does not set `HTTP_X_FORWARDED_PROTO`, you can add it as an allowed variable as well (step 5), and set its value to `https` (steps 9 & 10), assuming you use the reverse proxy to serve your BaGet over https.
